nntrainer/app_context.cpp

```diff
@@ -0,0 +1,85 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file	app_context.cpp
+ * @date	10 November 2020
+ * @brief	This file contains app context related functions and classes that
+ * manages the global configuration of the current environment
+ * @see		https://github.com/nnstreamer/nntrainer
+ * @author	Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug		No known bugs except for NYI items
+ *
+ */
+#include <dirent.h>
+#include <iostream>
+#include <sstream>
+
+#include <app_context.h>
+#include <nntrainer_log.h>
+#include <util_func.h>
+
+namespace nntrainer {
+
+/**
+ * @brief initiate global context
+ *
+ */
+static void init_global_context_nntrainer(void) __attribute__((constructor));
+
+/**
+ * @brief finialize global context
+ *
+ */
+static void fini_global_context_nntrainer(void) __attribute__((destructor));
+
+static void init_global_context_nntrainer(void) {}
+
+static void fini_global_context_nntrainer(void) {}
+
+AppContext &AppContext::Global() {
+  static AppContext instance;
```

Comment:
Any reason to not create a private member `static AppContext global` inside `AppContext` class itself?

Reply:
static local are guaranteed to be thread safe by standard. I am not sure global variable(static data member) is thread safe or not. For now, I couldn't think of the cases that I have to access in otherway. So I have gone to the static local variable

Reply:
static local members of a function of a class and static member of a class are both thread safe as long as there is 1 caller.
Two callers to the member function of a class modifying that thread variable will not be thread safe and will need synchronization primitives (not needed for now).
But this limits the static variable to that function. Later when other functionalities would also be provided, class static variable would be more versatile IMO.

Reply:
So this is the only one then?

Reply:
Lock will not protect against changing while network is running. If user changes while running, and its owned by the user, its on them IMO.

Reply:
I'll change this to static member

It's a bit different story but model will copy own the context to avoid temporal coupling. Changing context after setting model will not take effect. it won't be that expensive because context is a small class

Reply:
> So this is the only one then? I wonder that is it ok to change this in the middle of processing by other network? Maybe add some lock?

Local static member is guaranteed to be thread-safe from c++11

Reply:
> > So this is the only one then? I wonder that is it ok to change this in the middle of processing by other network? Maybe add some lock?
> 
> Local static member is guaranteed to be thread-safe from c++11

What I mean is that we shouldn't allow changing this in the middle of processing. That's what I mean. Stop and changing it and run is Ok. but changing it in the middle of processing, may cause problem I guess.

Reply:
The issue is handled in #733. Model copies the app_context which fixate the app context used for the model. So even if we change in the middle of the process, it remains safe.